source/developer/redux.rst

```diff
@@ -0,0 +1,261 @@
+          const {currentUserId} = getState().entities.users;
+          const preference = {
+              user_id: currentUserId,
+              category: Preferences.CATEGORY_FLAGGED_POST,
+              name: postId,
+              value: 'true'
+          };
+
+          savePreferences(currentUserId, [preference])(dispatch, getState);
+      };
+  }
+
+Make sure to also add your function to the default export at the bottom of the file.
+
+Testing the Action
+~~~~~~~~~~~~~~~~~~
+
+The final piece is testing your action. We use the `mochajs framework <https://mochajs.org/>`__ for testing, along with the `nock server mocking framework <https://github.com/node-nock/nock>`__ to mock the server where needed.
+
+The tests for actions live in ``test/actions/``. Add your test to the appropriate file following one of the many examples for the other actions.
+
+Make sure to read the `README <https://github.com/mattermost/mattermost-redux/blob/master/README.md>`__ for information on running the tests.
+
+
+Adding a Selector
+------------------
+
+Selectors are the method used to retrieving data from the state of the store. We use `reselect <https://github.com/reactjs/reselect>`__. If you'd like to know more about reselect and how we use it at Mattermost, `check out this developer talk given by core developer Harrison Healey <https://www.youtube.com/watch?v=6N2X7gEwmaQ>`__.
+
+Selectors must:
+- Receive ``state`` as the first argument and data returned must be solely based on what's in the state
+- Be created with ``createSelector`` whenever the data is manipulated or formatted before return
+- Be unit tested
+
+Selectors live in the ``src/selectors/`` directory.
+
+Implementing the Selector
+~~~~~~~~~~~~~~~~~~
+
+If your selector is just pulling data directly from the state without any manipulation, simply just return the data you need.
+
+
+.. code-block:: javascript
+  :linenos:
+
+  export function getUser(state, id) {
+      return state.entities.users.profiles[id];
+  }
+
+The above example is just simply pulling a user out of the profiles entity and requires no computation or formatting.
+
+If your selector needs to select based on some more advanced requirements or needs the result in a specific format then you'll need to make use of the ``createSelector`` function from `reselect <reselect <https://github.com/reactjs/reselect>`__. If you're not sure what this is good for, `check out the previously mentioned developer talk <https://www.youtube.com/watch?v=6N2X7gEwmaQ>`__. The short form reason is using reselect allows for memoization and only runs the computation of selectors when the state affecting that selector has actually changed.
+
+The basic usage for ``createSelector`` is pass it all the selector functions needed as inputs to your computation. The last argument will then be a function that takes in the results of each selector you specified and performs your computation on that data before returning it.
```

Comment:
"is pass" -> "is to pass"
"The last argument will then" -> "The last argument should then"
"each selector you specified and performs your computation on that data before returning it." -> "each previous selector, performs some computations, and then returns the result."